pyteal/compiler/subroutines.py

```diff
@@ -108,63 +110,90 @@ def spillLocalSlotsDuringRecursion(
                 # restore the local slots after returning from the subroutine. This prevents a
                 # reentry into the current subroutine from modifying variables we are currently
                 # using.
+
+                coverSpilledSlots = False
+                uncoverArgs = False
+                digArgs = True
+                if version >= Op.cover.min_version:
+                    digArgs = False
+                    if len(slots) < numArgs:
+                        coverSpilledSlots = True
+                    else:
+                        uncoverArgs = True
+
                 for slot in slots:
                     # spill local slots to the stack
                     before.append(TealOp(None, Op.load, slot))
 
+                    if coverSpilledSlots:
+                        # numArgs is guaranteed to be at least 2 here (since numArgs > len(slots)
+                        # and len(slots) must be at least 1 for the code to get this far), so no
+                        # need to replace this with swap if numArgs is 1
+                        before.append(TealOp(None, Op.cover, numArgs))
+
                 for _ in range(numArgs):
                     # pull the subroutine arguments to the top of the stack, above the just spilled
-                    # local slots
-
-                    # TODO: TEAL 5+, do this instead:
-                    # before.append(TealOp(None, Op.uncover, len(slots)))
-                    # or just do cover during the previous loop where slots are loaded, whichever
-                    # is more efficient
-                    before.append(
-                        TealOp(
-                            None,
-                            Op.dig,
-                            len(slots) + subroutine.argumentCount() - 1,
+                    # local slots, if needed
+
+                    stackDistance = len(slots) + numArgs - 1
+
+                    if uncoverArgs:
+                        if stackDistance == 1:
+                            before.append(TealOp(None, Op.swap))
+                        else:
+                            before.append(TealOp(None, Op.uncover, stackDistance))
+
+                    if digArgs:
+                        before.append(
+                            TealOp(
+                                None,
+                                Op.dig,
+                                stackDistance,
+                            )
                         )
-                    )
-                    # because we are stuck using dig instead of uncover in TEAL 4, we'll need to
-                    # pop all of the dug up arguments after the function returns
+                        # because we are stuck using dig instead of uncover in TEAL 4, we'll need to
+                        # pop all of the dug up arguments after the function returns
 
-                preserveReturnValue = False
+                coverAvailable = version >= Op.cover.min_version
```

Comment:
I think this logic condition can be put before line 117 (some if condition deciding if `cover` available), what do you think?

Reply:
Good idea